File: sdk-js/src/sign.js
import { canonicalize } from "./canonicalize.js";
import { base64UrlEncode, signCanonicalPayload } from "./crypto.js";
import { PayloadValidationError } from "./errors.js";

function stripSignature(payload) {
  const { sig, ...body } = payload;
  return body;
}

export function signPayload(payload, privateKeyPem) {
  if (!payload || typeof payload !== "object" || Array.isArray(payload)) {
    throw new PayloadValidationError("Payload must be a JSON object");
  }

  if (typeof payload.alg !== "string" || payload.alg.length === 0) {
    throw new PayloadValidationError("Missing alg");
  }

  const canonicalBody = canonicalize(stripSignature(payload));
  const signature = signCanonicalPayload(canonicalBody, privateKeyPem, payload.alg);

  return {
    ...stripSignature(payload),
    sig: base64UrlEncode(signature)
  };
}

File: sdk-python/document_trust_payload/sign.py
from .canonicalize import canonicalize
from .crypto import base64url_encode, sign_canonical_payload
from .errors import PayloadValidationError


def _strip_signature(payload):
    body = dict(payload)
    body.pop("sig", None)
    return body


def sign_payload(payload, private_key_pem: str):
    if not isinstance(payload, dict):
        raise PayloadValidationError("Payload must be a JSON object")

    algorithm = payload.get("alg")
    if not isinstance(algorithm, str) or not algorithm:
        raise PayloadValidationError("Missing alg")

    canonical_body = canonicalize(_strip_signature(payload))
    signature = sign_canonical_payload(canonical_body.encode("utf-8"), private_key_pem, algorithm)
    signed = _strip_signature(payload)
    signed["sig"] = base64url_encode(signature)
    return signed
